ocular/src/keyring.rs

```diff
@@ -0,0 +1,740 @@
+
+        Ok(PublicKeyOutput {
+            name: name.to_string(),
+            address,
+        })
+    }
+
+    fn get_all_keys(&self) -> Result<Vec<PublicKeyOutput>, Box<dyn std::error::Error>> {
+        let mut vec = Vec::new();
+
+        for entry in fs::read_dir(&self.key_path).expect("Could not read directory.") {
+            let path = entry.unwrap().path();
+
+            if path.is_file() {
+                if let Some(extension) = path.extension() {
+                    if extension == "pem" {
+                        let name = path
+                            .file_stem()
+                            .expect("Could not get file stem.")
+                            .to_str()
+                            .expect("Could not convert to string.");
+                        let key_data = self.get_key_address(name)?;
+
+                        vec.push(key_data);
+                    }
+                }
+            }
+        }
+
+        Ok(vec)
+    }
+
+    fn recover_from_mnemonic(
+        &self,
+        name: &str,
+        mnemonic: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError> {
+        // Check if key already exists
+        if self.key_exists(name)? && !override_if_exists {
+            eprintln!("Key '{}', already exists.", name);
+            return Err(KeyStoreError::Exists(name.to_string()));
+        }
+
+        let seed = Seed::new(
+            &bip39::Mnemonic::from_phrase(mnemonic.trim(), bip39::Language::English)
+                .expect("Could not read mnemonic."),
+            password,
+        );
+
+        let derivation_path = match derivation_path {
+            Some(_i) => derivation_path.unwrap(),
+            _ => COSMOS_BASE_DERIVATION_PATH,
+        };
+
+        let derivation_path = derivation_path
+            .parse::<bip32::DerivationPath>()
+            .expect("Could not parse derivation path.");
+
+        // Process key and store
+        let key =
+            bip32::XPrv::derive_from_path(seed, &derivation_path).expect("Could not derive key.");
+        let private_key = k256::SecretKey::from(key.private_key());
+        let encoded_key = private_key
+            .to_pkcs8_der()
+            .expect("Could not PKCS8 encode private key");
+
+        let key_name = &KeyName::new(name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", name));
+        self.key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .store(key_name, &encoded_key)
+            .expect("Could not store key");
+
+        Ok(())
+    }
+}
+
+// ---------------------------------- Tests ----------------------------------
+// TODO: Make these tests more comprehensive and increase code coverage.
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn file_key_store_without_path_init_test() {
```

Comment:
Nit: `_test` at the end of the method names is a little redundant IMO. Just a style preference that doesn't really matter.

Reply:
Cleaned up.